src/Http/SaloonRequest.php

```diff
@@ -3,156 +3,188 @@
-    {
-        if (! $this->loadedConnector instanceof SaloonConnector) {
-            $this->bootConnector();
-        }
-
-        return $this->loadedConnector;
-    }
-
-    /**
-     * Build up the final request URL.
-     *
-     * @return string
-     * @throws SaloonInvalidConnectorException
-     */
-    public function getFullRequestUrl(): string
-    {
-        $requestEndpoint = $this->defineEndpoint();
-
-        if ($requestEndpoint !== '/') {
-            $requestEndpoint = ltrim($requestEndpoint, '/ ');
-        }
-
-        $requiresTrailingSlash = ! empty($requestEndpoint) && $requestEndpoint !== '/';
-
-        $baseEndpoint = rtrim($this->getConnector()->defineBaseUrl(), '/ ');
-        $baseEndpoint = $requiresTrailingSlash ? $baseEndpoint . '/' : $baseEndpoint;
-
-        return $baseEndpoint . $requestEndpoint;
-    }
-
-    /**
-     * Check if a trait exists on the connector.
-     *
-     * @param string $trait
-     * @return bool
-     * @throws SaloonInvalidConnectorException
-     */
-    public function traitExistsOnConnector(string $trait): bool
-    {
-        return array_key_exists($trait, class_uses($this->getConnector()));
-    }
-
-    /**
-     * Dynamically proxy other methods to the connector.
-     *
-     * @param $method
-     * @param $parameters
-     * @return mixed
-     * @throws SaloonMethodNotFoundException
-     */
-    public function __call($method, $parameters)
-    {
-        if (method_exists($this->getConnector(), $method) === false) {
-            throw new SaloonMethodNotFoundException($method, $this->getConnector());
-        }
-
-        return $this->getConnector()->{$method}(...$parameters);
-    }
+	use CollectsData,
+		CollectsQueryParams,
+		CollectsHeaders,
+		CollectsConfig,
+		CollectsHandlers,
+		CollectsInterceptors,
+		SendsRequests;
+
+	/**
+	 * Define the method that the request will use.
+	 *
+	 * @var string|null
+	 */
+	protected ?string $method = null;
+
+	/**
+	 * Define a custom response that the request will return.
+	 *
+	 * @var string|null
+	 */
+    protected ?string $response = null;
+
+	/**
+	 * The connector.
+	 *
+	 * @var string|null
+	 */
+	protected ?string $connector = null;
+
+	/**
+	 * The instantiated connector instance.
+	 *
+	 * @var SaloonConnector|null
+	 */
+	private ?SaloonConnector $loadedConnector = null;
+
+	/**
+	 * Define anything to be added to the request.
+	 *
+	 * @return void
+	 */
+	public function boot(): void
+	{
+		// TODO: Implement boot() method.
+	}
+
+	/**
+	 * Get the method the class is using.
+	 *
+	 * @return string|null
+	 */
+	public function getMethod(): ?string
+	{
+		if(empty($this->method)) {
+			return null;
+		}
+
+		return $this->method;
+	}
+
+	/**
+	 * Get the response class
+	 * 
+	 * @return string
+	 * @throws \ReflectionException
+	 * @throws SaloonInvalidResponseClassException
+	 */
+	public function getResponseClass(): string
+	{
+		if(is_null($this->response)) return SaloonResponse::class;
```

Comment:
We should also probably check if a response exists on the connector, too? 
`$response = is_null($this->response) ? $this->getConnector()->getResponse() : $this->response`

Reply:
At first I thought of this same code, but checking if the "response" doesn't exist and already returning the default, you don't need to do unnecessary checks.

I will change for the suggested style and check about the connector.

personally I like the style:

```php
$response = $this->response ?? SaloonResponse::class;
```

What do you think?

```php
$response = $this->response ?? $this->getConnector()->getResponse()
```
Then the connector would return the default "SaloonResponse::class"

Reply:
Yes perfect 🤩

Reply:
Ok,

Changed, please note, 

I change the connector interface to contain getResponseClass method.
I create a HasResponseClass trait to use in SaloonConnector to implement getResponseClass.

Reply:
Next step is change HasCustomResponses to can be used in SaloonRequest too.

Maybe only changing to this:

```php
$response = $this->response;

if(!$response && method_exists($this, "getConnector")) {
	$response = $this->getConnector()->getResponseClass();
}
```

So this will work in Connector and Request.

Reply:
Ok, change the HasCustomResponses and I think only need to write test to connector.

Reply:
Added unit tests

Reply:
Could the method exists be replaced with $this instanceof SaloonRequest?

Reply:
Also do we even need reflectionclass in here anymore? Couldn’t we just check if $response instanceof SaloonResponse?